### ISSUE TITLE
reader_concurrency_semaphore: count resources are redundant

### ISSUE THREAD

**Issue by denesb** [scylladb/scylla#13547]
Initially, the `reader_concurrency_semaphore` was purely count based. Each read would consume exactly one point. This was wildly inaccurate because obviously reads have very different memory footprints. For some workloads, 100 concurrent reads is completely fine, for others, even 10 is too much.

To fix this, the semaphore was changed to hold memory instead of count resources. Each read would have a starting cost, then it would register all its memory usage on the semaphore. The semaphore would admit only so many reads, such that their collective memory consumption is lower than the limit. This quickly exploded because the initial memory accounting very inaccurate, reads often having real memory consumption order(s) of magnitude larger than the registered one. So we quickly backpedaled and made the semaphore restricted on both count and memory resources. The semaphore was still supposed to control concurrency mainly by memory, with the count-based admission control only being there as a fail-over "don't-go-crazy", in case the memory based concurrency control brakes down.

Since then, the accuracy of the memory tracking was vastly improved, to the point where I'm quite confident in its accuracy. We also moved the semaphore in front of the cache and introduced another admission constraint, where we only admit reads if no currently ongoing read demands only CPU to make progress. We also recently introduced hard upper limits on the memory consumption of reads.

All this has lead me to question whether we really still need count resources at all. On enterprise releases, each workload prioritization group has its own semaphore, with a portion of the total memory, proportional to its shares. All semaphores however has the original count resources of 100. So in theory creating many workload prioritization groups, or even just two of them, would create problems. The fact that we haven't seen a single instance where they do, strongly suggests, that the count resources are unneeded at this point and we can rely on just the memory.

Ditching the count resources would allow us to lift the mostly arbitrary count limits of 10 (for system and maintenance) and 100 (for statement groups). These hard limits don't scale with the size of nodes, unlike the memory limits which are expressed as a certain percentage of the shard's memory.

**Comment by asias** [comment]
Hope we could have it in the near future ;-)

**Comment by bhalevy** [comment]
The question is whether we need the count limit for small reads that don't need much memory but we still want to cap their parallelism.

**Comment by asias** [comment]
Also, as I mentioned here https://github.com/scylladb/scylla-enterprise/issues/3089#issuecomment-1628428561.      

Perhaps we can start with increasing the count limit from 10 to 100 (10X) for the streaming class.  This is less aggressive change than dropping the count limiter completely.

What do you think?

**Comment by bhalevy** [comment]
Increasing the limit for the maintenance semaphore makes sense as a first step.

That said, I hate the fact that the limit is constant.
@xemul can we come up with a way to derive the count limit based on the io setup?
Maybe it's worth doing as a follow up.

**Comment by denesb** [comment]
Since I opened this issue, I have seen numerous occurrences of semaphores having consumed all 100 of their count resources, but only a fraction of their memory resources. This can happen if reads mainly go to disk but don't consume a lot of memory (or there is a lot of memory available).

In conclusion I think it is not safe to just remove the count resources just yet. We need a mechanism to ensure disk concurrency is kept at a reasonable level. In lieu of that, we need to keep using the imperfect count resources.

The best would be if there would be a notification mechanism from seastar I/O, such that it would emit a signal when the disk queue for a certain scheduling group is saturated and another one when there is free capacity. Reads would be then queued if there is no disk capacity and admitted if there is free capacity, in addition to the existing CPU capacity admission criteria.

**Comment by mykaul** [comment]
> Since I opened this issue, I have seen numerous occurrences of semaphores having consumed all 100 of their count resources, but only a fraction of their memory resources. This can happen if reads mainly go to disk but don't consume a lot of memory (or there is a lot of memory available).

How can we/I easily identify this (and not ask you every time to take a look at the issue?)

**Comment by bhalevy** [comment]
Now with the improved i/o queue scheduling and rate limiting is the count-based concurrency limit really needed?

**Comment by denesb** [comment]
> > Since I opened this issue, I have seen numerous occurrences of semaphores having consumed all 100 of their count resources, but only a fraction of their memory resources. This can happen if reads mainly go to disk but don't consume a lot of memory (or there is a lot of memory available).
> 
> How can we/I easily identify this (and not ask you every time to take a look at the issue?)

The first line of the diagnostics dump will look something like this:

```
[shard  7] reader_concurrency_semaphore - (rate limiting dropped 3207 similar messages) Semaphore sl:default_read_concurrency_sem with 100/100 count and 4631647/170959831 memory resources: timed out, dumping permit diagnostics:
```

Note the `100/100 count` part. This means that all count resources were used up.

**Comment by denesb** [comment]
Also see https://github.com/scylladb/scylladb/blob/master/docs/dev/reader-concurrency-semaphore.md, this document has some guidelines on how to interpret these dumps. I will happily extend/clarify it if you have feedback.

**Comment by denesb** [comment]
> @xemul Now with the improved i/o queue scheduling and rate limiting is the count-based concurrency limit really needed?

We need either direct timeout support, or cancellation support from the I/O queue to be able to ditch the count resources. Otherwise it is possible that we flood the I/O queue with reads that will be timed out by the time they complete, and the replica will be doing nothing but wasting work.

**Comment by bhalevy** [comment]
> > @xemul Now with the improved i/o queue scheduling and rate limiting is the count-based concurrency limit really needed?
> 
> We need either direct timeout support, or cancellation support from the I/O queue to be able to ditch the count resources. Otherwise it is possible that we flood the I/O queue with reads that will be timed out by the time they complete, and the replica will be doing nothing but wasting work.

We already cancel I/Os when their associated input stream is closed.  This happens when the flat mutation reader times out and is destroyed.

What's the difference between timing out up to 100 admitted readers (and canceling their outstanding IOs) and doing the same for N >> 100?

In the latter case most IOs will still be queued in the scylla io_queues (and hence easily cancelable). The number of IOs sent to the disk that are currently non-cancelable won't change.

**Comment by avikivity** [comment]
High concurrency is only needed if we have a disk with high latency.

Assuming 20k ops/shard, concurrency = 100 requires a latency of 100/20k = 5ms. Which isn't so high. But we won't realistically achieve 20k ops/shard reading from sstables. 2k miss/shard = 4k reads/shard -> 25ms.

So, lifting the count limit looks to be beneficial for hard disks and S3, but not SSD.

For 4k reads/shard with 0.5ms latency, concurrency is just 2.

**Comment by denesb** [comment]
> High concurrency is only needed if we have a disk with high latency.
> 
> Assuming 20k ops/shard, concurrency = 100 requires a latency of 100/20k = 5ms. Which isn't so high. But we won't realistically achieve 20k ops/shard reading from sstables. 2k miss/shard = 4k reads/shard -> 25ms.
> 
> So, lifting the count limit looks to be beneficial for hard disks and S3, but not SSD.
> 
> For 4k reads/shard with 0.5ms latency, concurrency is just 2.

Doesn't the I/O queues already make sure we don't push more to the disk than what is beneficial? Or are you worried that the requests that actually get admitted to the disk are a mix of all the outstanding reads, and therefore all of them will proceed concurrently, just slower?

**Comment by avikivity** [comment]
> > High concurrency is only needed if we have a disk with high latency.
> > Assuming 20k ops/shard, concurrency = 100 requires a latency of 100/20k = 5ms. Which isn't so high. But we won't realistically achieve 20k ops/shard reading from sstables. 2k miss/shard = 4k reads/shard -> 25ms.
> > So, lifting the count limit looks to be beneficial for hard disks and S3, but not SSD.
> > For 4k reads/shard with 0.5ms latency, concurrency is just 2.
> 
> Doesn't the I/O queues already make sure we don't push more to the disk than what is beneficial? 

They do.

> Or are you worried that the requests that actually get admitted to the disk are a mix of all the outstanding reads, and therefore all of them will proceed concurrently, just slower?

Relaxing the 100 count limit introduces risk, that we misaccount something and run out of memory. What do we get in return for this risk?

For SSD, we get nothing.

For HDD, we also get nothing - the latency calculation indicates we do, but no real disk array will give us 4k IOPS/shard with a reasonable number of shards. In general good disk concurrency is 3X the number of disks in the array, and 2 shards already indicate 70 disks are needed.

For S3, we do get better throughput by increasing the concurrency.

**Comment by denesb** [comment]
> Relaxing the 100 count limit introduces risk, that we misaccount something and run out of memory. What do we get in return for this risk?

The benefits are:
* Much simpler code, dual-resource limiting is confusing and somewhat clunky.
* Could greatly help semaphores like the streaming semaphore, which today are severely limited concurrency wise (10 count units).

We have some limited evidence that the count based resource can be discarded safely. In Enterprise, each service-level has 100 new count resources "printed" for it, you can effectively increase the amount of disk concurrency by creating new service-levels. I'm not aware of any reports of this causing problems so far.

**Comment by mykaul** [comment]
You are limited by the amount of service levels (I think?)

**Comment by denesb** [comment]
> You are limited by the amount of service levels (I think?)

Yes, but hat is quite a generous limit, you can get 10x as many count resources easily, when compared to OSS for example.
